File: sdk/javascript/src/symbol/CryptoTypesV2.js
import { hexToUint8, uint8ToHex } from '../utils/converter.js';
import { CryptoScheme, PUBLIC_KEY_SIZES, PRIVATE_KEY_SIZES, SIGNATURE_SIZES } from './CryptoScheme.js';

/**
 * Variable-length byte array for V2 cryptographic values.
 * Unlike the V1 ByteArray (which enforces a fixed size), this allows
 * sizes determined at runtime by CryptoSchemeId.
 */
class VariableSizeByteArray {
	/**
	 * Creates a variable-size byte array.
	 * @param {Uint8Array|string} arrayInput Byte array or hex string.
	 * @param {string} name Type name for diagnostics.
	 */
	constructor(arrayInput, name = 'VariableSizeByteArray') {
		let rawBytes = arrayInput;
		if ('string' === typeof rawBytes)
			rawBytes = hexToUint8(rawBytes);

		this.bytes = new Uint8Array(rawBytes);
		this._name = name;
	}

	/** Returns the byte length. */
	get size() {
		return this.bytes.length;
	}

	toString() {
		return uint8ToHex(this.bytes);
	}

	toJson() {
		return this.toString();
	}
}

/**
 * Represents a V2 public key (variable-length: 32B for Ed25519, 1952B for ML-DSA-65).
 */
export class PublicKeyV2 extends VariableSizeByteArray {
	static NAME = 'PublicKeyV2';

	/**
	 * Creates a V2 public key.
	 * @param {Uint8Array|string} publicKey Public key bytes or hex string.
	 */
	constructor(publicKey) {
		super(publicKey instanceof PublicKeyV2 ? publicKey.bytes : publicKey, PublicKeyV2.NAME);
	}

	/**
	 * Infers the crypto scheme from this key's size.
	 * @returns {number} CryptoScheme identifier.
	 */
	get cryptoScheme() {
		for (const [scheme, size] of Object.entries(PUBLIC_KEY_SIZES)) {
			if (Number(size) === this.bytes.length) return Number(scheme);
		}
		throw new Error(`cannot infer crypto scheme from public key size ${this.bytes.length}`);
	}

	/**
	 * Creates a V2 public key validating against a crypto scheme.
	 * @param {number} schemeId Crypto scheme identifier.
	 * @param {Uint8Array|string} publicKey Public key bytes or hex.
	 * @returns {PublicKeyV2} Validated public key.
	 */
	static fromScheme(schemeId, publicKey) {
		const pk = new PublicKeyV2(publicKey);
		const expected = PUBLIC_KEY_SIZES[schemeId];
		if (undefined === expected) throw new Error(`unsupported crypto scheme: ${schemeId}`);
		if (pk.bytes.length !== expected)
			throw new RangeError(`public key size ${pk.bytes.length} does not match scheme ${schemeId} (expected ${expected})`);
		return pk;
	}
}

/**
 * Represents a V2 signature (variable-length: 64B for Ed25519, 3309B for ML-DSA-65).
 */
export class SignatureV2 extends VariableSizeByteArray {
	static NAME = 'SignatureV2';

	constructor(signature) {
		super(signature, SignatureV2.NAME);
	}

	get cryptoScheme() {
		for (const [scheme, size] of Object.entries(SIGNATURE_SIZES)) {
			if (Number(size) === this.bytes.length) return Number(scheme);
		}
		throw new Error(`cannot infer crypto scheme from signature size ${this.bytes.length}`);
	}

	static zero(schemeId = CryptoScheme.ED25519) {
		const size = SIGNATURE_SIZES[schemeId];
		if (undefined === size) throw new Error(`unsupported crypto scheme: ${schemeId}`);
		return new SignatureV2(new Uint8Array(size));
	}
}

/**
 * Represents a V2 private key (variable-length: 32B for Ed25519, 5984B for ML-DSA-65).
 */
export class PrivateKeyV2 extends VariableSizeByteArray {
	static NAME = 'PrivateKeyV2';

	constructor(privateKey) {
		super(privateKey, PrivateKeyV2.NAME);
	}

	get cryptoScheme() {
		for (const [scheme, size] of Object.entries(PRIVATE_KEY_SIZES)) {
			if (Number(size) === this.bytes.length) return Number(scheme);
		}
		throw new Error(`cannot infer crypto scheme from private key size ${this.bytes.length}`);
	}
}

export default {
	PublicKeyV2,
	SignatureV2,
	PrivateKeyV2,
	VariableSizeByteArray
};

File: sdk/python/symbolchain/symbol/CryptoTypesV2.py
"""Variable-length V2 cryptographic types for post-quantum crypto support."""

from binascii import hexlify, unhexlify

from .CryptoScheme import CryptoScheme, PUBLIC_KEY_SIZES, PRIVATE_KEY_SIZES, SIGNATURE_SIZES


class VariableSizeByteArray:
	"""Variable-length byte array for V2 cryptographic values."""

	def __init__(self, array_input, tag=None):
		"""Creates a variable-size byte array from bytes or hex string."""
		raw_bytes = array_input
		if isinstance(raw_bytes, str):
			raw_bytes = unhexlify(raw_bytes)
		if isinstance(raw_bytes, bytearray):
			raw_bytes = bytes(raw_bytes)

		self.bytes = raw_bytes
		self._tag = tag

	@property
	def size(self):
		"""Returns the byte length."""
		return len(self.bytes)

	def __eq__(self, other):
		return isinstance(other, VariableSizeByteArray) and self.bytes == other.bytes and self._tag == other._tag

	def __ne__(self, other):
		return not self == other

	def __hash__(self):
		return hash(self.bytes)

	def __str__(self):
		return hexlify(self.bytes).decode('utf8').upper()

	def __repr__(self):
		return f'{self._tag.__name__ if self._tag else "VariableSizeByteArray"}(\'{str(self)}\')'

	def to_json(self):
		"""Returns representation that can be stored in JSON."""
		return str(self)


class PublicKeyV2(VariableSizeByteArray):
	"""Represents a V2 public key (variable-length)."""

	def __init__(self, public_key):
		raw = public_key.bytes if isinstance(public_key, PublicKeyV2) else public_key
		super().__init__(raw, PublicKeyV2)

	@property
	def crypto_scheme(self):
		"""Infers the crypto scheme from this key's size."""
		for scheme, size in PUBLIC_KEY_SIZES.items():
			if size == len(self.bytes):
				return scheme
		raise ValueError(f'cannot infer crypto scheme from public key size {len(self.bytes)}')

	@staticmethod
	def from_scheme(scheme_id, public_key):
		"""Creates a V2 public key validating against a crypto scheme."""
		pk = PublicKeyV2(public_key)
		expected = PUBLIC_KEY_SIZES.get(scheme_id)
		if expected is None:
			raise ValueError(f'unsupported crypto scheme: {scheme_id}')
		if len(pk.bytes) != expected:
			raise ValueError(f'public key size {len(pk.bytes)} does not match scheme {scheme_id} (expected {expected})')
		return pk


class SignatureV2(VariableSizeByteArray):
	"""Represents a V2 signature (variable-length)."""

	def __init__(self, signature):
		super().__init__(signature, SignatureV2)

	@property
	def crypto_scheme(self):
		"""Infers the crypto scheme from this signature's size."""
		for scheme, size in SIGNATURE_SIZES.items():
			if size == len(self.bytes):
				return scheme
		raise ValueError(f'cannot infer crypto scheme from signature size {len(self.bytes)}')

	@staticmethod
	def zero(scheme_id=CryptoScheme.ED25519):
		"""Creates a zeroed signature for the given scheme."""
		size = SIGNATURE_SIZES.get(scheme_id)
		if size is None:
			raise ValueError(f'unsupported crypto scheme: {scheme_id}')
		return SignatureV2(bytes(size))


class PrivateKeyV2(VariableSizeByteArray):
	"""Represents a V2 private key (variable-length)."""

	def __init__(self, private_key):
		super().__init__(private_key, PrivateKeyV2)

	@property
	def crypto_scheme(self):
		"""Infers the crypto scheme from this key's size."""
		for scheme, size in PRIVATE_KEY_SIZES.items():
			if size == len(self.bytes):
				return scheme
		raise ValueError(f'cannot infer crypto scheme from private key size {len(self.bytes)}')
